src/coreclr/tools/aot/ILCompiler.Compiler/IL/ILImporter.Scanner.cs

```diff
@@ -346,6 +441,41 @@ private void ImportCall(ILOpcode opcode, int token)
                     Debug.Assert(false); break;
             }
 
+            // Are we scanning a call to an async method from an async state machine?
+            if (opcode is ILOpcode.call or ILOpcode.callvirt
+                && _canonMethod.IsAsyncCall()
+                && method.IsAsync)
```

Comment:
Is the `IsAsync` check correct for (abstract) virtual methods? They do not have an implementation so I would not expect us to consult the Impl flag to figure out how the call is going to be performed.

Reply:
Interesting - so interface calls would never get optimized to runtime async calling convention?

Reply:
> For a task-returning interface method, the impl flag does not introduce a lot of difference 

BTW, `async` abstract/interface methods are illegal in C#. 

The IL spec for Runetime Async says:

* The [MethodImpl(MethodImplOptions.Async)] only has effect when applied to method definitions that return generic or nongeneric variants of Task or ValueTask.
* The [MethodImpl(MethodImplOptions.Async)] only has effect when applied to method definitions with CIL implementation.

Basically - it is ok to have a `MethodImplOptions.Async`  on an abstract/interface task-returning method, but semantically it makes no difference. 

That matches the part where the VM/AOT/typesystem adds a pair of variant methods either way.